reciprocalspaceship/io/mtz.py

```diff
@@ -133,16 +133,16 @@ def to_gemmi(
     mtz.datasets[0].dataset_name = dataset_name
 
     # Construct data for Mtz object
-    temp = dataset.reset_index()
+    # GH#255: DataSet is provided using the range_indexed decorator
     columns = []
-    for c in temp.columns:
-        cseries = temp[c]
+    for c in dataset.columns:
+        cseries = dataset[c]
         if isinstance(cseries.dtype, MTZDtype):
             mtz.add_column(label=c, type=cseries.dtype.mtztype)
             columns.append(c)
         # Special case for CENTRIC and PARTIAL flags
         elif cseries.dtype.name == "bool" and c in ["CENTRIC", "PARTIAL"]:
-            temp[c] = temp[c].astype("MTZInt")
+            dataset[c] = dataset[c].astype("MTZInt")
```

Comment:
this line changes the dtype of the passed dataset's 'CENTRIC' column. can we avoid this side effect? one way would be to convert it back to bools after saving, but maybe there is a cleaner way? 

```
[ins] In [23]: ds = rs.read_mtz("unmerged.mtz")[['I', 'SIGI']].label_centrics()

[ins] In [24]: ds.dtypes
Out[24]:
I          Intensity
SIGI          Stddev
CENTRIC         bool
dtype: object

[ins] In [25]: ds.write_mtz('test.mtz')

[ins] In [26]: ds.dtypes
Out[26]:
I          Intensity
SIGI          Stddev
CENTRIC       MTZInt
PARTIAL         bool
dtype: object
```

Reply:
A new "PARTIAL" column is also created. I'd prefer it if this wasn't added. It's clear from the docstring there's something special about PARTIAL and CENTRIC, but I didn't expect my dataset to be modified by calling this method.

Reply:
That's a fair point, but if I'm not mistaken, these are pre-existing side-effects that are unrelated to this change, right? The CENTRIC/PARTIAL columns need to be converted to an MTZ compatible column type to be written out. The "PARTIAL" in your example gets added because it's an unmerged dataset -- that's also a requirement for M/ISYM handling in the MTZ file format. 

I agree that it would be nice to avoid any changes to the DataSet upon writing -- the simplest way to do that would be to copy the dataset prior to changes, but I'd rather avoid unnecessary memory/computation. Do you think these changes are best handled here? Or should we file an additional Issue and handle in the future?

Reply:
Ah, this may have been introduced as a side effect by this PR. Previously, `reset_index()` had been called without `inplace=True`, whereas it is now called inplace due to the `range_indexed` decorator. As such, in the past these operations were done on a copy of the calling dataset. 

As such, I think the simplest fix is to copy the calling dataset. It feels suboptimal, but it does preserve the previous behavior.

Reply:
Ok I think I fixed this in the latest set of commits